pallets/pallet-assets-handler/src/lib.rs

```diff
@@ -0,0 +1,233 @@
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Stores an asset id on chain under an associated resource ID.
+		#[pallet::call_index(0)]
+		#[pallet::weight({1000})]
+		pub fn set_resource(
+			origin: OriginFor<T>,
+			resource_id: ResourceId,
+			asset: AssetInfo<AssetId<T>, BalanceOf<T>>,
+		) -> DispatchResult {
+			<T as pallet_bridge::Config>::BridgeCommitteeOrigin::ensure_origin(origin)?;
+			ResourceToAssetInfo::<T>::insert(resource_id, asset.clone());
+			Self::deposit_event(Event::ResourceUpdated { resource_id, asset });
+			Ok(())
+		}
+
+		/// Removes a resource ID from the resource mapping.
+		///
+		/// After this call, bridge transfers with the associated resource ID will
+		/// be rejected.
+		#[pallet::call_index(1)]
+		#[pallet::weight({1000})]
+		pub fn remove_resource(origin: OriginFor<T>, resource_id: ResourceId) -> DispatchResult {
+			<T as pallet_bridge::Config>::BridgeCommitteeOrigin::ensure_origin(origin)?;
+			ResourceToAssetInfo::<T>::remove(resource_id);
+			Self::deposit_event(Event::ResourceRemoved { resource_id });
+			Ok(())
+		}
+	}
+}
+
+impl<T, B, A> BridgeHandler<B, A, ResourceId> for Pallet<T>
+where
+	T: Config
+		+ frame_system::Config<AccountId = A>
+		+ pallet_bridge::Config<Balance = B>
+		+ pallet_assets::Config<Balance = B>
+		+ pallet_balances::Config<Balance = B>,
+	B: Copy + FixedPointOperand + CheckedSub,
+	A: Clone,
+{
+	fn prepare_token_bridge_in(
+		resource_id: ResourceId,
+		who: A,
+		amount: B,
+	) -> Result<B, DispatchError> {
+		let asset_info = Self::resource_to_asset_info(resource_id);
+		match asset_info {
+			None => Err(Error::<T>::InvalidResourceId.into()),
+			// Native token
+			Some(AssetInfo { fee: _, asset: None }) => {
+				Self::deposit_event(Event::TokenBridgeIn {
+					asset_id: None,
+					to: who.clone(),
+					amount,
+				});
+				pallet_balances::Pallet::<T>::mint_into(&who, amount)
+			},
+			// pallet assets
+			Some(AssetInfo { fee: _, asset: Some(asset) }) => {
+				Self::deposit_event(Event::TokenBridgeIn {
+					asset_id: Some(asset.clone()),
+					to: who.clone(),
+					amount,
+				});
+				pallet_assets::Pallet::<T>::mint_into(asset, &who, amount)
+			},
+		}
+	}
+	// Return actual amount to target chain after deduction e.g fee
+	fn prepare_token_bridge_out(
+		resource_id: ResourceId,
+		who: A,
+		amount: B,
+	) -> Result<B, DispatchError> {
+		let asset_info = Self::resource_to_asset_info(resource_id);
+		match asset_info {
+			None => Err(Error::<T>::InvalidResourceId.into()),
+			// Native token
+			Some(AssetInfo { fee, asset: None }) => {
+				Self::deposit_event(Event::TokenBridgeOut {
+					asset_id: None,
+					to: who.clone(),
+					amount,
+					fee,
+				});
+				let burn_amount = pallet_balances::Pallet::<T>::burn_from(
+					&who,
+					amount,
+					Precision::Exact,
+					Fortitude::Polite,
+				)?;
+				ensure!(burn_amount > fee, Error::<T>::CannotPayAsFee);
+				pallet_balances::Pallet::<T>::mint_into(&T::TreasuryAccount::get(), fee)?;
```

Comment:
I feel the logic is not that comprehensive: why don't we simply burn `amount + fee` balance? 🤔

Reply:
There are two parts of logic mixing here.
(1) (burn_from and mint_into) is fungible trait method which is going to replace currency trait in future.
The precision and fortitude choice will allow the corresponding call change exact/not exact amount of token + whether the account should still exist after call.
Yes.. it looks complicated and old currency trait is easier.

(2) For compatibility of this version, I wish to have the exact same event as previous token bridge, in case the bridge transfer event with after fee/before fee amount is delicate.